### PR TITLE
begin ZD client integration, add index method

### PR DESCRIPTION
**Pull request by zacharyehren**
Hey Ben, I began the Rails API Client integration which can be found here. I was feeling a little rusty with Rails and want to make sure everything looks good so far. For reference, here's the client's documentation: https://github.com/zendesk/zendesk_api_client_rb

My notes/questions can be found below:

- I added the `zendesk.rb` file to initializers and added my ZD account's info to it. I saw there was an Oauth token field but commented that out. I don't think I'll need that for this but wanted to check with you if you agree. Also, the client's documentation said this `Note: This ZendeskAPI API client only supports basic authentication at the moment.` So I'm guessing judging from that note, oauth doesn't work yet on here anyway. Can you make sure everything else on this file looks good?

- I added an index method to the Tickets Controller and used the `client.tickets` command from the Client's [Collections ](https://github.com/zendesk/zendesk_api_client_rb/wiki/Collections )methods. I'm now trying to test the index method using the following curl command: `curl 'localhost:3000/api/tickets'` but receiving a long error message that begins with the following:

```
"status":404,"error":"Not Found","exception":"#\u003cActionController::RoutingError: uninitialized constant Api\u003e",
```
This brings up two questions:
1. Am I testing index correctly?
2. Any ideas why the error is happening? I'm thinking that I may need to do some kind of call/authorization with  the `Zen_Client` instance which I currently am attempting by adding `require ZEN_CLIENT` in the controller. This is most likely incorrect.

- Lastly I just want to make sure everything else looks good with how I'm setting this up - such as the route and the Ticket model & Tickets Controller

### CODE REVIEW

File: config/initializers/zendesk.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ZEN_CLIENT = ZendeskAPI::Client.new do |config|
+  # Mandatory:
+
+  config.url = "https://travelingyeti.zendesk.com" # e.g. https://mydesk.zendesk.com/api/v2
+
+  # Basic / Token Authentication
+  config.username = "thetravelingyeti@gmail.com"
+
+  # Choose one of the following depending on your authentication choice
+  config.token = "SJjEznBljjLwqI95HYSuhpFwPky80dDKB7y0noOl"
```
Comment (bmneely) [inline config/initializers/zendesk.rb]: You probably don't want to commit this to git or reveal it on GitHub.

File: app/controllers/tickets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+require ZEN_CLIENT
```
Comment (bmneely) [inline app/controllers/tickets_controller.rb]: You shouldn't need to require this. It should be available throughout the app.

File: app/controllers/tickets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+
+class TicketsController < ApplicationController
+
+  def index
+    @tickets = client.tickets
```
Comment (bmneely) [inline app/controllers/tickets_controller.rb]: `client` should be `ZEN_CLIENT`

**Review by bmneely — DISMISSED** [review]
You're getting the `RoutingError` because you don't have your controller setup to be under the API 
namespace. Look back on how you structured your controllers here: 

https://github.com/zacharyehren/todo-api/blob/master/app/controllers/api_controller.rb

and here: 
https://github.com/zacharyehren/todo-api/blob/master/app/controllers/api/items_controller.rb

**Comment by zacharyehren** [comment]
I'm assuming that moving 1 step forward and 2 steps back is normal when building an app? Because that's how this is feeling :) 

So now that I realized I wasn't calling my Rails API, I updated the URL on the Angular side. But from there I realized that the Rails API isn't working since the homepage is no longer rendering anything. To start, I'm just testing the index method and have found two scenarios happening:

## Scenario 1

When the index method is as follows:

```ruby
def index
    @tickets = ZEN_CLIENT.tickets
    puts @tickets
    render @tickets
  end
```
I receive the following output in the Rails server when I use `curl 'http://localhost:3000/api/tickets'`:

```
Started GET "/api/tickets" for 127.0.0.1 at 2017-10-16 18:55:28 -0700
Processing by Api::TicketsController#index as JSON
ticket collection [path=tickets]
Completed 200 OK in 7ms (Views: 0.7ms)
```
Which is a different output than what we saw last Thursday during our call. There should be a JSON output of all the tickets. Also, on the front-end, nothing is logged to the console. The only difference in the code since last Thursday is the use of the Figaro gem. So I went ahead and tested the curl command when the token is hardcoded instead of used through application.yml and received the same result. 

## Scenario 2

Just to see what would happen, I tested the following with the index method:

```ruby
def index
    @tickets = ZEN_CLIENT.tickets
    puts @tickets
    render json: @tickets
  end
```
And received the following output from the Rails server:

```
Started GET "/api/tickets" for 127.0.0.1 at 2017-10-16 18:47:26 -0700
Processing by Api::TicketsController#index as JSON
ticket collection [path=tickets]
I, [2017-10-16T18:47:26.817064 #18425]  INFO -- : get https://travelingyeti.zendesk.com/api/v2/tickets
D, [2017-10-16T18:47:26.817180 #18425] DEBUG -- :   Accept: "application/json"
  Accept-Encoding: "gzip;q=1.0,deflate;q=0.6,identity;q=0.3"
  User-Agent: "ZendeskAPI Ruby 1.15.0"
I, [2017-10-16T18:47:27.730290 #18425]  INFO -- : Status 401
D, [2017-10-16T18:47:27.730787 #18425] DEBUG -- :   server: "nginx"
  date: "Tue, 17 Oct 2017 01:47:27 GMT"
  content-type: "application/json; charset=UTF-8"
  content-length: "37"
  connection: "close"
  www-authenticate: "Basic realm=\"Web Password\""
  strict-transport-security: "max-age=31536000;"
  x-ua-compatible: "IE=Edge,chrome=1"
  cache-control: "no-cache, private"
  x-zendesk-origin-server: "app38.pod14.use1.zdsys.com"
  x-request-id: "768ea778-1a01-42cf-c2ae-126ce9f57dc6"
  x-runtime: "0.051379"
  x-rack-cache: "miss"
  x-zendesk-request-id: "feab56e44898fb64d3dd"
{"error"=>"Couldn't authenticate you"}
[active_model_serializers] Rendered ActiveModel::Serializer::CollectionSerializer with ActiveModelSerializers::Adapter::Attributes (0.39ms)
Completed 200 OK in 920ms (Views: 918.0ms)
```

The main thing I noticed was `"error"=>"Couldn't authenticate you"`. And then on the front-end, an empty array is logged to the console.

So all that said, I'm not sure which scenario I should be focusing on. Since scenario 1 was working fine on Thursday, I'm having trouble finding what has changed since then. I also find scenario 2 troubling since everything should be fine from an authentication standpoint. 

Since I'm not worried about anyone hacking my Zendesk trial account, here's the token I'm using to help with the troubleshooting:

`SJjEznBljjLwqI95HYSuhpFwPky80dDKB7y0noOl`

**Comment by zacharyehren** [comment]
Hey Ben, just wanted to check to see if you've had a chance to look at my comment above

**Comment by bmneely** [comment]
Hmm, do you see the same issue if you hit the ZenDesk API directly (not going through the FE or Rails app)?

**Comment by zacharyehren** [comment]
No it was working swimmingly when I was sending requests to the ZD API directly. I had the tickets listed on the homepage and was able to click into the individual tickets as well.

**Comment by zacharyehren** [comment]
Sorry, I may have misunderstood. My previous response was referring to making the calls directly through Angular. Meaning:

```javascript
  var commentInfo = {
          method: 'GET',
          url: 'https://travelingyeti.zendesk.com/api/v2/tickets/' + $cookies.get('zendeskTicketId') + '/comments',
          headers: {
            'Authorization': 'Basic ' + window.btoa(client.username + '/token:' + client.token)
          }
        }
```

When I try and call it directly with a curl command, I'm seeing the following: 

screen shot 2017-10-18 at 11 31 00 pm

**Comment by bmneely** [comment]
Calling it directly with curl will still require you to send the auth header. However, if it was working correctly via Angular, and is not working via Rails, then it is most likely something with Rails. Do you have your latest (post figaro) changes pushed up to your Rails API repo?

**Comment by zacharyehren** [comment]
Yep! These are the most up-to-date files for the app.

**Comment by bmneely** [comment]
I think I found it, let me know if changing that fixes the issue. You will
have to restart your rails app after any change to an initializer.

-- 
Ben Neely

On October 19, 2017 at 12:42:26 PM, zacharyehren (notifications@github.com)
wrote:

Yep! These are the most up-to-date files for the app.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub
<https://github.com/zacharyehren/zendesk-api/pull/1#issuecomment-338000126>,
or mute the thread

**Comment by zacharyehren** [comment]
Hey Ben, apologies for the radio silence the last few days. I had family in town over the weekend. 

I started to get really tired of the Rails Client that we've been trying to work with so I decided I'm going to call Zendesk directly with my Rails API. To do this, I followed how we did it for the Kele project and created a Class called `ZD_Request` (found in zd_requests.rb). The main difference between this and Kele is that I decided to use the 'rest-client' gem instead of 'http-party'. 

I then created a method called `get_comments` which sends a `get` request for the comments in the ticket (comments are the initial ticket body + the follow-up responses). 

What I'd like to do from here is use this method in the `Show` action in the tickets controller. This brings up a couple of questions:

1. Am I approaching this the correct way? I think calling ZD directly will be easier than completely relying on this funky Rails Client we've been trying to work with, but want to make sure that creating a Ruby Class is the right way to connect to the ZD endpoints. 
2. If I am approaching this correctly, how do I pull in the Class methods into the Tickets Controller? This is how I'm currently trying to do it:

```ruby
def show
    ticket = ZD_Request.get_comments(:id)
  end
```
I feel like this is basic Ruby I'm getting stuck on but feeling a little rusty.

**Comment by bmneely** [comment]
Building your own client isn't an approach I'd recommend (you're going end up spending a lot of time implementing the features that you get for free using the Zendesk API Client, I can absolutely guarantee you that it will ultimately be more work and more time to re-implement it yourself vs using an existing solution).

That being said, if it an approach you'd like to follow, you'll want to put it into `lib` instead of in an `initializers`.

Here are a couple resources on using `lib`:

* [What to put in lib](http://blog.codeclimate.com/blog/2012/02/07/what-code-goes-in-the-lib-directory/)
* [How to include lib files in your app](https://stackoverflow.com/questions/12724531/rails-engines-when-to-put-code-in-app-when-in-lib-and-when-in-vendor-folder)

**Comment by zacharyehren** [comment]
Ah ok. I was thinking about utilizing both. I have the `index` method working correctly with their client, I figure I'll keep using it for that. But I was having a really hard time figuring out how to access the `comments` object which would be utilized in the `show` method. Since I figured out how to access it by calling ZD directly back when I had it set up on Angular's end, I'm thinking I could use the same set up in Rails. 

Let's discuss more tomorrow! I'd love to solely use their client if possible. Really all that's left in terms of using accessing and creating data is displaying comments, creating new comments, and creating new tickets.

**Pull request merged**